src/cargo/core/compiler/job_queue.rs

```diff
@@ -36,13 +86,49 @@ pub struct JobQueue<'a, 'cfg> {
     queue: DependencyQueue<Unit<'a>, Artifact, Job>,
     tx: Sender<Message>,
     rx: Receiver<Message>,
-    active: HashMap<u32, Unit<'a>>,
+    active: HashMap<JobId, Unit<'a>>,
     compiled: HashSet<PackageId>,
     documented: HashSet<PackageId>,
     counts: HashMap<PackageId, usize>,
     progress: Progress<'cfg>,
     next_id: u32,
     timings: Timings<'a, 'cfg>,
+
+    /// These are tokens that we have acquried but aren't actively using.
+    ///
+    /// We should only have tokens here "in-between" loops of the
+    /// `drain_the_queue` function, i.e., when we've acquired a bunch of tokens
+    /// and are about to give them to rustcs.
+    unused_tokens: Vec<Acquired>,
```

Comment:
Oh sorry, to clarify, can this be a separate struct which only lives for the one stack frame where we're running stuff?

Reply:
Also, as to this field specifically, I don't think we're going to want it. There should never be any long-term storage for unused tokens.